src/app.js

```diff
@@ -1,26 +1,71 @@
 const getJobs = require("./getJobs");
+const getToken = require("./getToken");
+const getCookie = require("./getCookie");
 const express = require("express");
 const app = express();
-const port = 8081;
+const config = require("./config");
+const { host, port, AppID, APP_SECRET, redirect_url } = config;
 const path = require("path");
 
 app.use(express.static(path.join(__dirname)));
+app.set("views", __dirname + "/views");
 
 app.set("view engine", "ejs");
 
-//app.get("redirect", (req,res) => res.render("pages/redirect"))
+app.get("/redirect/:state", (req, res) => {
+  res.redirect(
+    `https://gitlab.com/oauth/authorize?` +
+      `&client_id=${AppID}` +
+      `&redirect_uri=${host}/redirect` +
+      `&response_type=code` +
+      `&state=${req.params.state}` +
+      `&scope=profile+read_api`
+  );
+});
 
-app.get("/groups/:id/jobs", function (req, res) {
-  getJobs.getJobs(req.params.id).then((data) =>
-    res.render("pages/index", {
-      created: data.filter((data) => data.status === "created"),
-      pending: data.filter((data) => data.status === "pending"),
-      running: data.filter((data) => data.status === "running"),
+app.get("/redirect", (req, res) => {
+  code = req.query.code;
+  state = req.query.state;
+  getToken
+    .getToken(AppID, code, APP_SECRET, redirect_url)
+    .then((response) => {
+      res.cookie(`access_token`, response.access_token, { maxAge: 86400000 });
+      res.redirect(`${host}/${state}`);
     })
-  );
-  // .then((data) => console.log((data[0]).map(({status, ...rest})=> [status[1], rest])));
+    .catch((err) =>
+      res.render("pages/error", {
+        error:
+          `Cannot obtain access token. ` +
+          `Please make sure Application ID, Application Secret, Redirect URL are valid`,
+      })
+    );
+});
+
+app.get("/groups/:id/jobs", function (req, res) {
+  const Token = getCookie.getCookie(req);
+  if (!Token) res.redirect(`${host}/redirect/${req.originalURL}`);
```

Comment:
It needs to be returned here, otherwise, the code will keep going after the redirect response.